### PR TITLE
Week 2: Comments

### PR DESCRIPTION
**Pull request by pr3pony**

### CODE REVIEW

**Review by shadyof — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,14 +19,60 @@
...
+
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: final

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,14 +19,60 @@
...
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: simplify as
return value == null ? defaultValue : value;

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -19,14 +19,60 @@
...
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: an interesting question for String:
String A = "";
then A == null ? true or false

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -19,14 +19,60 @@
...
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: try to use "final" if the variable shouldn't be assign anymore.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -19,14 +19,60 @@
...
+    PreparedQuery results = datastore.prepare(query);
+    ArrayList<String> comments = new ArrayList<String>();
+    for (Entity entity : results.asIterable()) {
+      comments.add((String) entity.getProperty("text"));
+    }
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: although it should depends on JVM, but usually C style "(int i = 0; i < x; i++)" for loop could faster than forEach style iterator

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (pr3pony) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: false

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (pr3pony) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think forEach is more readable than for-loop here

Comment (pr3pony) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: > To operate on large result sets, you should prefer asIterable(com.google.appengine.api.datastore.FetchOptions) and asIterator(com.google.appengine.api.datastore.FetchOptions), which stream the results from the datastore. [reference](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery#aslist)

So I think PreparedQuery.asIterable() is more general than PreparedQuery.asList() + for-loop

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -19,14 +19,60 @@
...
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private DatastoreService datastore;
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: some naming rule, I already forgot where the rules start but I know the rules it's being there for a long time.
FYI
https://source.android.com/setup/contribute/code-style#follow-field-naming-conventions
The rules can help you and others members figure out the potential meaning more quickly.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -19,14 +19,60 @@
...
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+    ArrayList<String> comments = new ArrayList<String>();
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: We can create a member in this class
private final ArrayList<String> mComments = new ArrayList<String>();
and try to re-use the member here, so we don't have to allocate memory so often.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -19,14 +19,60 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String text = getParameter(request, "text-input", "");
+    long timestamp = System.currentTimeMillis();
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("text", text);
+    commentEntity.setProperty("timestamp", timestamp);
+    datastore.put(commentEntity);
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: be careful about code injection attach, a simple solution is to filter some string before we read it( for example we can do this in #getParameter), but I think we can leave it now.

Comment (pr3pony) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: in my JS I put comments in innerText of p tag
so it should prevent XSS
many attempts failed [in the live version](http://jzhang-sps-summer20.appspot.com/index.html)

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -17,7 +17,16 @@
  */
 function addRandomGreeting() {
   const greetings =
-      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!'];
+      ["Pop Pop!", "Leonard likes this post.", "I can't believe I spent ten bucks on this! I don't know any of these people!", "My name is Alex!",
+      "We're the only species on earth that observes shark week.", "TV's the best dad there is. TV never came home drunk, TV never forgot me at the zoo, TV never abused and insulted me... unless you count Cop Rock.",
```
Comment (shadyof) [inline portfolio/src/main/webapp/script.js]: maybe change line here for better viewing.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -17,7 +17,16 @@
...
+      "The funny thing about being smart, is that you can get through most of life without ever having to do any work.",
+      "I see your value now.",
+      "Our first assignment is a documentary. The're like real movies, but with ugly people.",
+      "I looked inside Nicolas Cage and I found a secret: People are random and pointless.",
+      "I painted a tunnel on the side of the library. When it dries, I'm going for it.",
+      "Do you understand what a conspiracy is? When you conspire with everyone you come across, you're not really conspiring with anyone. You're just doing random crap.",
```
Comment (shadyof) [inline portfolio/src/main/webapp/script.js]: ditto

**Review by shadyof — COMMENTED** [review]
leave some comments

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (shadyof) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: okay cool!

**Pull request merged**